### ISSUE TITLE
Pickling Error

### ISSUE THREAD

**Issue by mrendu** [locie/pySIP#7]
Hello,
I get the following error when I run the "fit_bayesian" file: 
```
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "Xx\pySIP-0.8\fit_bayesian.py", line 43, in <module>
    fit = reg.fit(df=df, inputs=inputs, outputs='T_int')
  File "Xxx\anaconda3\envs\pysip\lib\site-packages\pysip\regressors\bayesian.py", line 124, in fit
    uchains, stats, options = dhmc.sample(q0, n_draws, n_chains, n_warmup, options)
  File "Xxx\anaconda3\envs\pysip\lib\site-packages\pysip\mcmc\hmc.py", line 193, in sample
    for n in range(n_chains)
  File "Xxx\anaconda3\envs\pysip\lib\site-packages\joblib\parallel.py", line 1017, in __call__
    self.retrieve()
  File "Xxx\anaconda3\envs\pysip\lib\site-packages\joblib\parallel.py", line 909, in retrieve
    self._output.extend(job.get(timeout=self.timeout))
  File "Xxx\anaconda3\envs\pysip\lib\site-packages\joblib\_parallel_backends.py", line 562, in wrap_future_result
    return future.result(timeout=timeout)
  File "Xxx\anaconda3\envs\pysip\lib\concurrent\futures\_base.py", line 435, in result
    return self.__get_result()
  File "Xxx\anaconda3\envs\pysip\lib\concurrent\futures\_base.py", line 384, in __get_result
    raise self._exception
PicklingError: Could not pickle the task to send it to the workers.
```
Do you have any idea what the problem is?
The "fit_frequentist" file runs properly.
Thanks

**Comment by Sarah-Ju** [comment]
Hi Manon,

I've had many times that same exact error :-(
it has to do with the fact that the bayesian runs multiple processes, one for each chain (4 by default) and somehow it does not work from ipython. There seems to be also numpy functions in the hamiltonian leaps that also use by default multithreading/processing (don't know the difference), but anyway, it doesn't help...

while waiting for a fix in the code, a few tricks made it work though:
- be sure to run on a python 3.7 environment
- not to run it on an ipython terminal, but from a normal terminal (make sure you activated your python work env)

Have you had some luck with that ?

Cheers

**Comment by jonasced** [comment]
Hi, 

This is quite late, but I tried your tricks (Python 3.7.12) and still received the same error in hmc.py 
```

Chain 0:   0%|                                                                                                                     | 0/2000 [00:00<?, ?it/sjoblib.externals.loky.process_executor._RemoteTraceback:                                                                            | 0/2000 [00:00<?, ?it/s]
"""in 2:   0%|                                                                                                                     | 0/2000 [00:00<?, ?it/s]
Traceback (most recent call last):                                                                                                 | 0/2000 [00:00<?, ?it/s]
  File "/home/jonas/anaconda3/envs/pysip/lib/python3.7/site-packages/joblib/externals/loky/backend/queues.py", line 153, in _feed
    obj_ = dumps(obj, reducers=reducers)
  File "/home/jonas/anaconda3/envs/pysip/lib/python3.7/site-packages/joblib/externals/loky/backend/reduction.py", line 271, in dumps
    dump(obj, buf, reducers=reducers, protocol=protocol)
  File "/home/jonas/anaconda3/envs/pysip/lib/python3.7/site-packages/joblib/externals/loky/backend/reduction.py", line 264, in dump
    _LokyPickler(file, reducers=reducers, protocol=protocol).dump(obj)
  File "/home/jonas/anaconda3/envs/pysip/lib/python3.7/site-packages/joblib/externals/cloudpickle/cloudpickle_fast.py", line 602, in dump
    return Pickler.dump(self, obj)

...

File "/home/jonas/anaconda3/envs/pysip/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
_pickle.PicklingError: Could not pickle the task to send it to the workers.

```

I tried looking around in the files resulting in the error. When opening the cloudpickle_fast.py file, I found this comment at the top:

```

"""
New, fast version of the CloudPickler.

This new CloudPickler class can now extend the fast C Pickler instead of the
previous Python implementation of the Pickler class. Because this functionality
is only available for Python versions 3.8+, a lot of backward-compatibility
code is also removed.

Note that the C Pickler subclassing API is CPython-specific. Therefore, some
guards present in cloudpickle.py that were written to handle PyPy specificities
are not present in cloudpickle_fast.py
"""
```

which seems to imply that it is not compatible with Python 3.7 or earlier versions. When trying fit_bayesian.py in a 3.8 environment I got a different error which seems to stem from "AttributeError: 'tqdm_asyncio' object has no attribute 'disable' ".
```
Traceback (most recent call last): 
  File "/home/jonas/.local/lib/python3.8/site-packages/tqdm/std.py", line 1147, in __del__                                                                
    self.close()
  File "/home/jonas/.local/lib/python3.8/site-packages/tqdm/std.py", line 1266, in close
    if self.disable:
AttributeError: 'tqdm_asyncio' object has no attribute 'disable'
Exception ignored in: <function tqdm.__del__ at 0x7fca1b2843a0>
Traceback (most recent call last):
  File "/home/jonas/.local/lib/python3.8/site-packages/tqdm/std.py", line 1147, in __del__
    self.close()
  File "/home/jonas/.local/lib/python3.8/site-packages/tqdm/std.py", line 1266, in close
...
joblib.externals.loky.process_executor._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/home/jonas/anaconda3/envs/test/lib/python3.8/site-packages/joblib-1.1.0-py3.8.egg/joblib/externals/loky/process_executor.py", line 407, in _process_worker
    call_item = call_queue.get(block=True, timeout=timeout)
  File "/home/jonas/anaconda3/envs/test/lib/python3.8/multiprocessing/queues.py", line 116, in get
    return _ForkingPickler.loads(res)
  File "/home/jonas/.local/lib/python3.8/site-packages/tqdm/utils.py", line 89, in __getattr__
    return getattr(self._wrapped, name)
  File "/home/jonas/.local/lib/python3.8/site-packages/tqdm/utils.py", line 89, in __getattr__
    return getattr(self._wrapped, name)
  File "/home/jonas/.local/lib/python3.8/site-packages/tqdm/utils.py", line 89, in __getattr__
    return getattr(self._wrapped, name)
  [Previous line repeated 988 more times]
RecursionError: maximum recursion depth exceeded
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/jonas/Local/pySIP/examples/thermal_network/armadillo/fit_bayesian.py", line 43, in <module>
    fit = reg.fit(df=df, inputs=inputs, outputs='T_int')
  File "/home/jonas/anaconda3/envs/test/lib/python3.8/site-packages/pysip-0.8-py3.8.egg/pysip/regressors/bayesian.py", line 123, in fit
  File "/home/jonas/anaconda3/envs/test/lib/python3.8/site-packages/pysip-0.8-py3.8.egg/pysip/mcmc/hmc.py", line 191, in sample
  File "/home/jonas/anaconda3/envs/test/lib/python3.8/site-packages/joblib-1.1.0-py3.8.egg/joblib/parallel.py", line 1056, in __call__
    self.retrieve()
  File "/home/jonas/anaconda3/envs/test/lib/python3.8/site-packages/joblib-1.1.0-py3.8.egg/joblib/parallel.py", line 935, in retrieve
    self._output.extend(job.get(timeout=self.timeout))
  File "/home/jonas/anaconda3/envs/test/lib/python3.8/site-packages/joblib-1.1.0-py3.8.egg/joblib/_parallel_backends.py", line 542, in wrap_future_result
    return future.result(timeout=timeout)
  File "/home/jonas/anaconda3/envs/test/lib/python3.8/concurrent/futures/_base.py", line 444, in result
    return self.__get_result()
  File "/home/jonas/anaconda3/envs/test/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception
joblib.externals.loky.process_executor.BrokenProcessPool: A task has failed to un-serialize. Please ensure that the arguments of the function are all picklable.
```

So, something else breaks, but I believe it could still be Python and its compatibility with Cloudpickle causing the trouble when running in 3.7. 

I've also tried setting n_cpu=1 in hmc.py/sample() manually with Python 3.7 to bypass the multithreading, and then I can run fit_bayesian.py but it is of course slower (though it seems to use 3/8 cores even without the code for multithreading? I'm also no expert in parallel programming so I find that a bit strange...).

**Comment by celliern** [comment]
Hello

I will close this issue as the rewrite of the package for the next pysip version will make it irrelevant : the bayesian part of the package is now dealt using pymc and should work seamlessly

Feel free to reopen-it after the new version release (or open a new issue if needed)
